chain/consensus/mir/state_manager.go

```diff
@@ -0,0 +1,309 @@
+package mir
+
+import (
+	"bytes"
+	"fmt"
+
+	"github.com/multiformats/go-multiaddr"
+	"go4.org/sort"
+
+	"github.com/filecoin-project/lotus/chain/consensus/hierarchical"
+	availabilityevents "github.com/filecoin-project/mir/pkg/availability/events"
+	"github.com/filecoin-project/mir/pkg/events"
+	"github.com/filecoin-project/mir/pkg/modules"
+	"github.com/filecoin-project/mir/pkg/pb/availabilitypb"
+	"github.com/filecoin-project/mir/pkg/pb/commonpb"
+	"github.com/filecoin-project/mir/pkg/pb/contextstorepb"
+	"github.com/filecoin-project/mir/pkg/pb/eventpb"
+	"github.com/filecoin-project/mir/pkg/pb/requestpb"
+	t "github.com/filecoin-project/mir/pkg/types"
+)
+
+const (
+	availabilityModuleID = t.ModuleID("availability")
+)
+
+type Message []byte
+
+type Batch struct {
+	Messages   []Message
+	Validators []t.NodeID
+}
+
+type StateManager struct {
+
+	// The current epoch number.
+	currentEpoch t.EpochNr
+
+	// For each epoch number, stores the corresponding membership.
+	memberships map[t.EpochNr]map[t.NodeID]t.NodeAddress
```

Comment:
Why do we need to store the history of memberships for every epoch?

Reply:
I see, you use it just to keep the latest two epochs, you don't keep the whole history.

Reply:
Yes, there is a garbage collection for old memberships.